v2/contribute/decisions/thirdparty/0001-showing-thirdparty-providers-list-dynamically.mdx

```diff
@@ -0,0 +1,106 @@
+
+** 4. Dynamic config and multi tenant **
+
+This would be a classic multi-tenant scenario, where each tenant has a different set of thirdparty providers enabled, and different credentials for them.
+
+### Problem Statement
+
+What would be the best way for the user to differentiate & configure the SDK for all the above use cases?
+
+## Considered Options
+
+* Show providers list dynamically when Multitenancy recipe is initialized
+* Always try to show providers list dynamically, fallback to statically configured list
+* Add `usesDynamicLoginMethods` config to the top level config
+
+## Decision Outcome
+
+Chosen option: **Add `usesDynamicLoginMethods` config to the top level config**, because
+
+* Clearly indicates that the dynamic providers list is being used
+* Avoids call to backend if the user does not intend to use dynamic providers list
+
+Also, `usesDynamicLoginMethods` is added as a toplevel config instead of thirdparty recipe, as this functionality will be expanded to other recipes in future.
+
+### This is how the user would configure the frontend in the mentioned use cases
+
+** Single Tenant, static config **
+
+No change in the config. `usesDynamicLoginMethods` is optional & `false` by default.
+
+** Multi Tenant, static config **
+
+Frontend is initialized Multitenancy.init() with a config to get tenantId.
+
+** Single Tenant, dynamic config **
+
+Frontend is initialized with `usesDynamicLoginMethods` set to `true`.
+
+** Multi Tenant, dynamic config **
+
+Frontend is initialized with `usesDynamicLoginMethods` set to `true`, and Multitenancy.init() with a config to get tenantId.
+
+## Pros and Cons of the Options
+
+### Show providers list dynamically when Multitenancy recipe is initialized
+
+This means that whenever the login methods and thirdparty providers list has to be dynamically fetched from the backend, Multitenancy recipe must be initialized.
+
+<ArgumentCon> From the user's point of view, it is unintuitive to call `Multitenancy.init()` for a single tenant / dynamic config use case </ArgumentCon>
+
+### Always try to show providers list dynamically, fallback to statically configured list
+
+Here, we always try to fetch the dynamic config from the backend. If the user has not configured in the core, the SDK would fallback to use the static configuration
+
+<ArgumentPro> No extra config needed by the user </ArgumentPro>
+<ArgumentCon> Ends up calling the backend API unnecessarily if the static provider list is enough for the user </ArgumentCon>
+
+### Add `usesDynamicLoginMethods` config to the thirdparty recipe
+
+<ArgumentPro> Backend API will be called only when the feature is enabled </ArgumentPro>
+<ArgumentPro> Clearly indicates that the dynamic providers list is being used </ArgumentPro>
+<ArgumentCon> In case of multi-tenancy, even though it's most common to use dynamic provider list, user has to enable this to avail that functionality </ArgumentCon>
```

Comment:
What about the option of setting this config to true by default if the user has initialised multi tenancy.init AND has provided their own impl of getTenantId? In that case, most probably, the user is using multi tenancy and then even if they do not set this boolean, things will just work?

That being said, i do not prefer this option cause it's harder to communicate to the user about whats happening, and harder for us to implement

Reply:
Done